torchmultimodal/models/coca/coca_model.py

```diff
@@ -0,0 +1,503 @@
+from torchmultimodal.modules.losses.contrastive_loss_with_temperature import (
+    ContrastiveLossWithTemperature,
+)
+
+
+class CoCaModelOutput(NamedTuple):
+    image_pooled_output: Tensor
+    text_pooled_output: Tensor
+    multimodal_embeddings: Tensor
+
+
+class CoCaModel(nn.Module):
+    """
+    CoCa model class containing vision encoder, text decoder, and multimodal decoder.
+    Reference: https://arxiv.org/abs/2205.01917
+    Args:
+        vision_encoder (nn.Module): Instantiated vision encoder. Should return either
+            TransformerOutput or Tensor.
+        text_decoder (CoCaTextDecoder): Instantiated CoCaTextDecoder returning a
+            Tuple[Tensor, Tensor], where the first element is the normalized CLS
+            embedding, and the second element is the full set of token embeddings.
+        multimodal_decoder (nn.Module): Instantiated CoCaMultimodalDecoder returning a
+            Tensor of multimodal embeddings.
+        vision_pooler (nn.Module): Pooler for vision outputs (see e.g. AttentionPooler).
+        vision_proj (nn.Module): Projection layer for vision encoder. Note that the
+            projections for the text_decoder and multimodal_decoder are handled inside
+            the CoCaTextDecoder and CoCaMultimodalDecoder classes, respectively, but
+            for vision we apply attentional pooling first so the vision projection
+            is separated from the vision_encoder class.
+    """
+
+    def __init__(
+        self,
+        vision_encoder: nn.Module,  # e.g. ViT
+        text_decoder: CoCaTextDecoder,
+        multimodal_decoder: CoCaMultimodalDecoder,
+        vision_pooler: nn.Module,
+        vision_proj: nn.Module,
+    ):
+        super().__init__()
+        self.vision_encoder = vision_encoder
+        self.text_decoder = text_decoder
+        self.multimodal_decoder = multimodal_decoder
+        self.vision_pooler = vision_pooler
+        self.vision_proj = vision_proj
+
+    def forward(
+        self, images: Tensor, texts: Tensor, text_padding_mask: Optional[Tensor] = None
+    ) -> CoCaModelOutput:
+        """
+        Args:
+            images (Tensor): Tensor of size (bsz, c, h, w) containing image pixels.
+            texts (Tensor): Tensor of size (bsz, seq_len) containing text tokens.
+            text_padding_mask (Optional[Tensor]): Boolean mask indicating padded tokens.
+                True for unpadded tokens, False for padded tokens. Default: None
+        Returns:
+            CoCaModelOutput containing pooled image embeddings, text embeddings,
+                and multimodal embeddings.
+        """
+
+        # Image encoder
+        vision_encoder_outs = self.vision_encoder(images)
+
+        # Extract image embeddings
+        if isinstance(vision_encoder_outs, TransformerOutput):
+            image_embeddings = vision_encoder_outs.last_hidden_state
+        elif isinstance(vision_encoder_outs, tuple):
+            vision_encoder_outs = vision_encoder_outs[0]
```

Comment:
This seems very risky. In general isinstance is very risky, but here especially since the user doesn't know without looking at the code that it's assumed that output 0 is the only thing being used.

Reply:
Yeah fair point. This is because a common return type from transformer encoders is a tuple (last_hidden_state, all_hidden_states). Maybe we could add an assert here? Something like 
```suggestion
            assert len(vision_encoder_outs) == 2 and isinstance(vision_encoder_outs[0], Tensor) and isinstance(vision_encoder_outs[1], List)
            vision_encoder_outs = vision_encoder_outs[0]
```